mephisto/core/task_launcher.py

```diff
@@ -83,13 +96,46 @@ def create_assignments(self) -> None:
                 )
                 self.units.append(Unit(self.db, unit_id))
 
-    def launch_units(self, url: str) -> None:
-        """launch any units registered by this TaskLauncher"""
         for unit in self.units:
+            self.unlaunched_units[unit.db_id] = unit
+
+    def generate_units(self):
+        """ units generator which checks that only 'max_num_concurrent_units' running at the same time,
+        i.e. in the LAUNCHED or ASSIGNED states """
+        while True:
+            for _, launched_unit in self.launched_units.copy().items():
+                if (
+                    launched_unit.get_status() != AssignmentState.LAUNCHED
+                    and launched_unit.get_status() != AssignmentState.ASSIGNED
+                ):
+                    self.num_running_units -= 1
+                    self.launched_units.pop(launched_unit.db_id)
+
+            num_avail_units = self.max_num_concurrent_units - self.num_running_units
+            for i, item in enumerate(self.unlaunched_units.copy().items()):
+                db_id, unit = item
+                if i < num_avail_units:
+                    self.launched_units[unit.db_id] = unit
+                    self.unlaunched_units.pop(unit.db_id)
+                    self.num_running_units += 1
+                    yield unit
```

Comment:
This is definitely functional, but rather than incurring the cost of a dict copy, you could instead bound your loop on `range(num_avail_units)` and select that many units. Alternately you could have a break in the `else` condition of the given if (instead of continuing the loop).

Ultimately though, this code is simplified by replacing the `unlaunched_units` dict with a list. We don't have an explicit need to directly index specific units from the list of units currently unlaunched, and as such it's a bag that we can take any value out of. Using a `list` then lets you simply use `unlaunched_units.pop()` for `num_avail_units` times.